### ISSUE TITLE
Break Code Into Features

### ISSUE THREAD

**Issue by cycle-five** [cycle-five/cracktunes#198]
In Rust, you can separate code into  "features" which can be turned on or off at compile time. This can greatly reduce the binary size and the length of the builds for situations where not everything is needed.

**Comment by 13unk0wn** [comment]
I am a bit confused as this is my first time contributing. Are you asking to identify the parts of the code that are not essential and wrap them in features so that others can include or exclude them as needed?

**Comment by cycle-five** [comment]
Exactly. I'll update this with more details later today, but some big ones are tracing, performance metrics (which still needs to be added), and separation of the modules which is rough at the moment and needs a lot more testing.
Sent from Proton Mail Android
-------- Original Message --------On 1/7/25 2:29 AM, 13unk0wn  wrote:
I am a bit confused as this is my first time contributing. Are you asking to identify the parts of the code that are not essential and wrap them in features so that others can include or exclude them as needed?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were assigned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/cycle-five/cracktunes/issues/198#issuecomment-2574564973",
"url": "https://github.com/cycle-five/cracktunes/issues/198#issuecomment-2574564973",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by 13unk0wn** [comment]
Okay, I will do my best to solve what I can.

**Comment by cycle-five** [comment]
Done — reclosing as completed rather than not-planned.

The workspace is split across feature flags: `crack-music`, `crack-playlist`, `crack-osint`, `crack-gpt`, `crack-bf`, `crack-voting`, `crack-testing`, `crack-types`, `crack-activity`, `crack-metrics`, `crack-telemetry`, `crack-tracing`, `cache`. The default build excludes osint, gpt, bf, voting and metrics entirely.

v0.4.1 finished the job by repairing the two that had rotted into not compiling: `crack-metrics` declared no `dep:prometheus` and its module was commented out of `lib.rs`, and `crack-telemetry` declared no dependencies at all while still calling into APIs whose imports had been commented out. Both build now, and `cargo clippy --all-features` is clean.
